File: src/numbers.cpp
#include "numbers.hpp"

#include <Arduino.h>

#include "pins.hpp"
#include "util.hpp"

bool firstNumberState = false;
bool secondNumberState = false;
bool thirdNumberState = false;
bool fourthNumberState = false;

void setNumber(bool a, bool b, bool c, bool d, bool e, bool f, bool g)
{
	digitalWriteBool(A, a);
	digitalWriteBool(B, b);
	digitalWriteBool(C, c);
	digitalWriteBool(D, d);
	digitalWriteBool(E, e);
	digitalWriteBool(F, f);
	digitalWriteBool(G, g);
}

void allOn()
{
	setNumber(true, true, true, true, true, true, true);
}

void zero()
{
    setNumber(true, true, true, true, true, true, false);
}

void one()
{
    setNumber(false, true, true, false, false, false, false);
}

void two()
{
    setNumber(true, true, false, true, true, false, true);
}

void three()
{
    setNumber(true, true, true, true, false, false, true);
}

void four()
{
    setNumber(false, true, true, false, false, true, true);
}

void five()
{
    setNumber(true, false, true, true, false, true, true);
}

void six()
{
    setNumber(true, false, true, true, true, true, true);
}

void seven()
{
	setNumber(true, false, false, false, true, true, true);
}

void eight()
{
    setNumber(true, true, true, true, true, true, true);
}

void nine()
{
    setNumber(true, true, true, true, false, true, true);
}

void setNumberParam(int number)
{
	switch (number)
	{
		case 1:
			one();
			break;
		case 2:
			two();
			break;
		case 3:
			three();
			break;
		case 4:
			four();
			break;
		case 5:
			five();
			break;
		case 6:
			six();
			break;
		case 7:
			seven();
			break;
		case 8:
			eight();
			break;
		case 9:
			nine();
			break;
		default:
			zero();
	}
}

void switchFirstNumber()
{
	firstNumberState = !firstNumberState;
	digitalWriteBool(H1, firstNumberState);
}

void switchSecondNumber()
{
	secondNumberState = !secondNumberState;
	digitalWriteBool(H2, secondNumberState);
}

void switchThirdNumber()
{
	thirdNumberState = !thirdNumberState;
	digitalWriteBool(M1, thirdNumberState);
}

void switchFourthNumber()
{
	fourthNumberState = !fourthNumberState;
	digitalWriteBool(M2, fourthNumberState);
}

File: node_api/numbers.ts
import { Gpio } from "onoff";
import { High, Low } from "./types";

export class Numbers
{
    H1: any;
    H2: any;
    M1: any;
    M2: any;
    A: any;
    B: any;
    C: any;
    D: any;
    E: any;
    F: any;
    G: any;

    firstNumberState: boolean = false;
    secondNumberState: boolean = false;
    thirdNumberState: boolean = false;
    fourthNumberState: boolean = false;

    constructor(H1: number, H2: number, M1: number, M2: number, 
        A: number, B: number, C: number, D: number, E: number, F: number, G: number)
    {
        if (H1 == 0 || H2 == 0 || M1 == 0 || M2 == 0 || 
            A == 0 || B == 0 || C == 0 || D == 0 || E == 0 || F == 0 || G == 0)
            return;
        this.H1 = new Gpio(H1, 'out');
        this.H2 = new Gpio(H2, 'out');
        this.M1 = new Gpio(M1, 'out');
        this.M2 = new Gpio(M2, 'out');
        this.A = new Gpio(A, 'out');
        this.B = new Gpio(B, 'out');
        this.C = new Gpio(C, 'out');
        this.D = new Gpio(D, 'out');
        this.E = new Gpio(E, 'out');
        this.F = new Gpio(F, 'out');
        this.G = new Gpio(G, 'out');

        this.firstNumberState = false;
        this.secondNumberState = false;
        this.thirdNumberState = false;
        this.fourthNumberState = false;

        this.switchFirstNumber(true);
        this.switchSecondNumber(true);
        this.switchThirdNumber(true);
        this.switchFourthNumber(true);
    }

    setNumber(a: boolean, b: boolean, c: boolean, d: boolean, e: boolean, f: boolean, g: boolean)
    {
        this.A.writeSync(a ? High : Low);
        this.B.writeSync(b ? High : Low);
        this.C.writeSync(c ? High : Low);
        this.D.writeSync(d ? High : Low);
        this.E.writeSync(e ? High : Low);
        this.F.writeSync(f ? High : Low);
        this.G.writeSync(g ? High : Low);
    }

    allOn()
    {
        this.setNumber(true, true, true, true, true, true, true);
    }

    zero()
    {
        this.setNumber(true, true, true, true, true, true, false);
    }

    one()
    {
        this.setNumber(false, true, true, false, false, false, false);
    }

    two()
    {
        this.setNumber(true, true, false, true, true, false, true);
    }

    three()
    {
        this.setNumber(true, true, true, true, false, false, true);
    }

    four()
    {
        this.setNumber(false, true, true, false, false, true, true);
    }

    five()
    {
        this.setNumber(true, false, true, true, false, true, true);
    }

    six()
    {
        this.setNumber(true, false, true, true, true, true, true);
    }

    seven()
    {
        this.setNumber(true, true, true, false, false, false, false);
    }

    eight()
    {
        this.setNumber(true, true, true, true, true, true, true);
    }

    nine()
    {
        this.setNumber(true, true, true, true, false, true, true);
    }

    setNumberParam(number: number)
    {
        switch (number)
        {
            case 1:
                this.one();
                break;
            case 2:
                this.two();
                break;
            case 3:
                this.three();
                break;
            case 4:
                this.four();
                break;
            case 5:
                this.five();
                break;
            case 6:
                this.six();
                break;
            case 7:
                this.seven();
                break;
            case 8:
                this.eight();
                break;
            case 9:
                this.nine();
                break;
            default:
                this.zero();
        }
    }

    switchNumbers()
    {
        if (!this.firstNumberState && !this.secondNumberState && !this.thirdNumberState && !this.fourthNumberState)
            this.switchFirstNumber();
        else
        {
            if (this.firstNumberState)
            {
                this.switchFirstNumber();
                this.switchSecondNumber();
            }
            else if (this.secondNumberState)
            {
                this.switchSecondNumber();
                this.switchThirdNumber();
            }
            else if (this.thirdNumberState)
            {
                this.switchThirdNumber();
                this.switchFourthNumber();
            }
            else if (this.fourthNumberState)
            {
                this.switchFourthNumber();
                this.switchFirstNumber();
            }
        }
    }

    switchFirstNumber(setState: boolean = false)
    {
        this.H1.writeSync(setState ? High : Low);
    }

    switchSecondNumber(setState: boolean = false)
    {
        this.H2.writeSync(setState ? High : Low);
    }

    switchThirdNumber(setState: boolean = false)
    {
        this.M1.writeSync(setState ? High : Low);
    }

    switchFourthNumber(setState: boolean = false)
    {
        this.M2.writeSync(setState ? High : Low);
    }
}